File: sdk/kotlin/src/main/kotlin/ai/meshllm/Node.kt
package ai.meshllm

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.callbackFlow
import kotlinx.coroutines.withContext
import uniffi.mesh_ffi.ChatMessageNative
import uniffi.mesh_ffi.ChatRequestNative
import uniffi.mesh_ffi.ClientEvent
import uniffi.mesh_ffi.ClientStatus
import uniffi.mesh_ffi.EventListener as FfiEventListener
import uniffi.mesh_ffi.MeshNodeHandleInterface
import uniffi.mesh_ffi.ModelNative
import uniffi.mesh_ffi.ResponsesRequestNative
import uniffi.mesh_ffi.createAutoNode as ffiCreateAutoNode
import uniffi.mesh_ffi.createNode as ffiCreateNode
import uniffi.mesh_ffi.discoverPublicMeshes as ffiDiscoverPublicMeshes

typealias CapabilityLevel = uniffi.mesh_ffi.CapabilityLevel
typealias CleanupPolicy = uniffi.mesh_ffi.CleanupPolicy
typealias CleanupResult = uniffi.mesh_ffi.CleanupResult
typealias DeleteModelOptions = uniffi.mesh_ffi.DeleteModelOptions
typealias DeleteModelResult = uniffi.mesh_ffi.DeleteModelResult
typealias DevicePolicy = uniffi.mesh_ffi.DevicePolicy
typealias DownloadedModel = uniffi.mesh_ffi.DownloadedModel
typealias InstalledModel = uniffi.mesh_ffi.InstalledModel
typealias LoadModelOptions = uniffi.mesh_ffi.LoadModelOptions
typealias ModelCacheStatus = uniffi.mesh_ffi.ModelCacheStatus
typealias ModelCapabilities = uniffi.mesh_ffi.ModelCapabilities
typealias ModelDetails = uniffi.mesh_ffi.ModelDetails
typealias ModelKind = uniffi.mesh_ffi.ModelKind
typealias ModelSearchQuery = uniffi.mesh_ffi.ModelSearchQuery
typealias ModelSource = uniffi.mesh_ffi.ModelSource
typealias ModelSummary = uniffi.mesh_ffi.ModelSummary
typealias PrunePolicy = uniffi.mesh_ffi.PrunePolicy
typealias PruneResult = uniffi.mesh_ffi.PruneResult
typealias PublicMesh = uniffi.mesh_ffi.PublicMesh
typealias PublicMeshQuery = uniffi.mesh_ffi.PublicMeshQuery
typealias ServedModel = uniffi.mesh_ffi.ServedModel
typealias ServingModelState = uniffi.mesh_ffi.ServingModelState
typealias ServingStatus = uniffi.mesh_ffi.ServingStatus
typealias UnloadModelOptions = uniffi.mesh_ffi.UnloadModelOptions
typealias UnloadTarget = uniffi.mesh_ffi.UnloadTarget

@JvmInline
value class InviteToken(val value: String)

data class Model(val id: String, val name: String)

data class ChatMessage(val role: String, val content: String)

data class ChatRequest(val model: String, val messages: List<ChatMessage>)

data class ResponsesRequest(val model: String, val input: String)

data class Status(val connected: Boolean, val peerCount: ULong)

@JvmInline
value class RequestId(val value: String)

sealed class Event {
    object Connecting : Event()
    data class Joined(val nodeId: String) : Event()
    data class ModelsUpdated(val models: List<Model>) : Event()
    data class TokenDelta(val requestId: RequestId, val delta: String) : Event()
    data class Completed(val requestId: RequestId) : Event()
    data class Failed(val requestId: RequestId, val error: String) : Event()
    data class Disconnected(val reason: String) : Event()
}

fun interface EventListener {
    fun onEvent(event: Event)
}

class Node(private val handle: MeshNodeHandleInterface) {
    val inference = Inference(handle)
    val models = Models(handle)
    val serving = Serving(handle)

    constructor(
        inviteToken: InviteToken,
        ownerKeypairBytesHex: String,
        cacheDir: String? = null,
        runtimeDir: String? = null,
        servingEnabled: Boolean = false,
    ) : this(ffiCreateNode(ownerKeypairBytesHex, inviteToken.value, cacheDir, runtimeDir, servingEnabled))

    suspend fun start(): Unit = withContext(Dispatchers.IO) { handle.start() }

    suspend fun stop(): Unit = withContext(Dispatchers.IO) { handle.stop() }

    suspend fun reconnect(): Unit = withContext(Dispatchers.IO) { handle.reconnect() }

    suspend fun status(): Status = withContext(Dispatchers.IO) { handle.status().toStatus() }

    class Inference(private val handle: MeshNodeHandleInterface) {
        suspend fun listModels(): List<Model> =
            withContext(Dispatchers.IO) { handle.inferenceListModels().map { it.toModel() } }

        fun chat(request: ChatRequest, listener: EventListener): RequestId {
            val bridge = object : FfiEventListener {
                override fun onEvent(event: ClientEvent) = listener.onEvent(event.toEvent())
            }
            return RequestId(handle.chat(request.toNative(), bridge))
        }

        fun responses(request: ResponsesRequest, listener: EventListener): RequestId {
            val bridge = object : FfiEventListener {
                override fun onEvent(event: ClientEvent) = listener.onEvent(event.toEvent())
            }
            return RequestId(handle.responses(request.toNative(), bridge))
        }

        fun cancel(requestId: RequestId) = handle.cancel(requestId.value)

        fun chatFlow(request: ChatRequest): Flow<Event> = callbackFlow {
            var requestId: RequestId? = null
            var terminalRequestId: RequestId? = null
            requestId = chat(request) { event ->
                trySend(event)
                val currentRequestId = requestId
                if (currentRequestId != null && event.isTerminalFor(currentRequestId)) {
                    terminalRequestId = currentRequestId
                    close()
                } else if (currentRequestId == null) {
                    terminalRequestId = event.terminalRequestId()
                }
            }
            if (requestId == terminalRequestId) {
                close()
            }
            awaitClose {
                val currentRequestId = requestId ?: return@awaitClose
                if (terminalRequestId != currentRequestId) {
                    cancel(currentRequestId)
                }
            }
        }

        fun responsesFlow(request: ResponsesRequest): Flow<Event> = callbackFlow {
            var requestId: RequestId? = null
            var terminalRequestId: RequestId? = null
            requestId = responses(request) { event ->
                trySend(event)
                val currentRequestId = requestId
                if (currentRequestId != null && event.isTerminalFor(currentRequestId)) {
                    terminalRequestId = currentRequestId
                    close()
                } else if (currentRequestId == null) {
                    terminalRequestId = event.terminalRequestId()
                }
            }
            if (requestId == terminalRequestId) {
                close()
            }
            awaitClose {
                val currentRequestId = requestId ?: return@awaitClose
                if (terminalRequestId != currentRequestId) {
                    cancel(currentRequestId)
                }
            }
        }
    }

    class Models(private val handle: MeshNodeHandleInterface) {
        suspend fun recommended(): List<ModelSummary> =
            withContext(Dispatchers.IO) { handle.recommendedModels() }

        suspend fun search(query: ModelSearchQuery): List<ModelSummary> =
            withContext(Dispatchers.IO) { handle.searchModels(query) }

        suspend fun show(modelRef: String): ModelDetails =
            withContext(Dispatchers.IO) { handle.showModel(modelRef) }

        suspend fun installed(): List<InstalledModel> =
            withContext(Dispatchers.IO) { handle.installedModels() }

        suspend fun cacheStatus(): ModelCacheStatus =
            withContext(Dispatchers.IO) { handle.modelCacheStatus() }

        suspend fun download(modelRef: String): DownloadedModel =
            withContext(Dispatchers.IO) { handle.downloadModel(modelRef) }

        suspend fun delete(modelRef: String, options: DeleteModelOptions): DeleteModelResult =
            withContext(Dispatchers.IO) { handle.deleteModel(modelRef, options) }

        suspend fun cleanup(policy: CleanupPolicy): CleanupResult =
            withContext(Dispatchers.IO) { handle.cleanupModels(policy) }

        suspend fun pruneDerivedCache(policy: PrunePolicy): PruneResult =
            withContext(Dispatchers.IO) { handle.pruneDerivedCache(policy) }
    }

    class Serving(private val handle: MeshNodeHandleInterface) {
        suspend fun status(): ServingStatus =
            withContext(Dispatchers.IO) { handle.servingStatus() }

        suspend fun servedModels(): List<ServedModel> =
            withContext(Dispatchers.IO) { handle.servedModels() }

        suspend fun load(modelRef: String, options: LoadModelOptions): ServedModel =
            withContext(Dispatchers.IO) { handle.loadServingModel(modelRef, options) }

        suspend fun unload(target: UnloadTarget, options: UnloadModelOptions): Unit =
            withContext(Dispatchers.IO) { handle.unloadServingModel(target, options) }

        suspend fun unloadModel(modelId: String, options: UnloadModelOptions): Unit =
            withContext(Dispatchers.IO) { handle.unloadServingModelById(modelId, options) }

        suspend fun unloadInstance(instanceId: String, options: UnloadModelOptions): Unit =
            withContext(Dispatchers.IO) { handle.unloadServingInstance(instanceId, options) }

        suspend fun setDevicePolicy(policy: DevicePolicy): Unit =
            withContext(Dispatchers.IO) { handle.setDevicePolicy(policy) }
    }

    companion object {
        suspend fun discoverPublicMeshes(
            query: PublicMeshQuery = PublicMeshQuery(
                model = null,
                minVramGb = null,
                region = null,
                targetName = null,
                relays = emptyList(),
            ),
        ): List<PublicMesh> = withContext(Dispatchers.IO) {
            ffiDiscoverPublicMeshes(query)
        }

        suspend fun connectPublic(
            ownerKeypairBytesHex: String,
            query: PublicMeshQuery = PublicMeshQuery(
                model = null,
                minVramGb = null,
                region = null,
                targetName = null,
                relays = emptyList(),
            ),
        ): Node = withContext(Dispatchers.IO) {
            Node(ffiCreateAutoNode(ownerKeypairBytesHex, query))
        }
    }
}

private fun ModelNative.toModel() = Model(id = id, name = name)

private fun ClientStatus.toStatus() = Status(connected = connected, peerCount = peerCount)

private fun ChatMessage.toNative() = ChatMessageNative(role = role, content = content)

private fun ChatRequest.toNative() =
    ChatRequestNative(model = model, messages = messages.map { it.toNative() })

private fun ResponsesRequest.toNative() = ResponsesRequestNative(model = model, input = input)

private fun ClientEvent.toEvent(): Event =
    when (this) {
        is ClientEvent.Connecting -> Event.Connecting
        is ClientEvent.Joined -> Event.Joined(nodeId = nodeId)
        is ClientEvent.ModelsUpdated -> Event.ModelsUpdated(models = models.map { it.toModel() })
        is ClientEvent.TokenDelta -> Event.TokenDelta(requestId = RequestId(requestId), delta = delta)
        is ClientEvent.Completed -> Event.Completed(requestId = RequestId(requestId))
        is ClientEvent.Failed -> Event.Failed(requestId = RequestId(requestId), error = error)
        is ClientEvent.Disconnected -> Event.Disconnected(reason = reason)
    }

private fun Event.isTerminalFor(requestId: RequestId): Boolean =
    when (this) {
        is Event.Completed -> this.requestId == requestId
        is Event.Failed -> this.requestId == requestId
        else -> false
    }

private fun Event.terminalRequestId(): RequestId? =
    when (this) {
        is Event.Completed -> requestId
        is Event.Failed -> requestId
        else -> null
    }

File: sdk/swift/Sources/MeshLLM/Node.swift
import Foundation

public struct InviteToken: Sendable {
    public let value: String

    public init(_ value: String) {
        self.value = value
    }
}

public struct Model: Sendable {
    public let id: String
    public let name: String
}

public struct Status: Sendable {
    public let connected: Bool
    public let peerCount: Int
}

public struct RequestId: Sendable {
    public let value: String
}

public enum Event: Sendable {
    case connecting
    case joined(nodeId: String)
    case modelsUpdated(models: [Model])
    case tokenDelta(requestId: String, delta: String)
    case completed(requestId: String)
    case failed(requestId: String, error: String)
    case disconnected(reason: String)
}

public struct ChatMessage: Sendable {
    public let role: String
    public let content: String

    public init(role: String, content: String) {
        self.role = role
        self.content = content
    }
}

public struct ChatRequest: Sendable {
    public let model: String
    public let messages: [ChatMessage]

    public init(model: String, messages: [ChatMessage]) {
        self.model = model
        self.messages = messages
    }
}

public struct ResponsesRequest: Sendable {
    public let model: String
    public let input: String

    public init(model: String, input: String) {
        self.model = model
        self.input = input
    }
}

#if canImport(MeshLLMFFI)
public final class Node: @unchecked Sendable {
    private let handle: MeshNodeHandle

    public let inference: Inference
    public let models: Models
    public let serving: Serving

    public init(
        inviteToken: InviteToken,
        ownerKeypairBytesHex: String,
        cacheDir: String? = nil,
        runtimeDir: String? = nil,
        servingEnabled: Bool = false
    ) throws {
        let handle = try createNode(
            ownerKeypairBytesHex: ownerKeypairBytesHex,
            inviteToken: inviteToken.value,
            cacheDir: cacheDir,
            runtimeDir: runtimeDir,
            servingEnabled: servingEnabled
        )
        self.handle = handle
        self.inference = Inference(handle: handle)
        self.models = Models(handle: handle)
        self.serving = Serving(handle: handle)
    }

    public init(handle: MeshNodeHandle) {
        self.handle = handle
        self.inference = Inference(handle: handle)
        self.models = Models(handle: handle)
        self.serving = Serving(handle: handle)
    }

    public static func discoverPublicMeshes(
        _ query: PublicMeshQuery = PublicMeshQuery(
            model: nil,
            minVramGb: nil,
            region: nil,
            targetName: nil,
            relays: []
        )
    ) async throws -> [PublicMesh] {
        try await runBlocking {
            try MeshLLM.discoverPublicMeshes(query: query)
        }
    }

    public static func connectPublic(
        ownerKeypairBytesHex: String,
        query: PublicMeshQuery = PublicMeshQuery(
            model: nil,
            minVramGb: nil,
            region: nil,
            targetName: nil,
            relays: []
        )
    ) async throws -> Node {
        let handle = try await runBlocking {
            try createAutoNode(ownerKeypairBytesHex: ownerKeypairBytesHex, query: query)
        }
        return Node(handle: handle)
    }

    public func start() async throws {
        let handle = self.handle
        try await runBlocking {
            try handle.start()
        }
    }

    public func stop() async throws {
        let handle = self.handle
        try await runBlocking {
            try handle.stop()
        }
    }

    public func reconnect() async throws {
        let handle = self.handle
        try await runBlocking {
            try handle.reconnect()
        }
    }

    public func status() async -> Status {
        let handle = self.handle
        let status = await runBlocking {
            handle.status()
        }
        return Status(connected: status.connected, peerCount: Int(clamping: status.peerCount))
    }

    public final class Inference: @unchecked Sendable {
        private let handle: MeshNodeHandle

        fileprivate init(handle: MeshNodeHandle) {
            self.handle = handle
        }

        public func listModels() async throws -> [Model] {
            let handle = self.handle
            let models = try await runBlocking {
                try handle.inferenceListModels()
            }
            return models.map(Node.mapModel)
        }

        public func chat(_ request: ChatRequest) -> AsyncThrowingStream<Event, Error> {
            let native = Node.mapChatRequest(request)
            return AsyncThrowingStream { continuation in
                do {
                    let bridge = EventStreamBridge(continuation: continuation) { [handle] requestId in
                        try? handle.cancel(requestId: requestId)
                    }
                    let requestId = try handle.chat(request: native, listener: bridge)
                    bridge.activate(requestId: requestId)
                } catch {
                    continuation.finish(throwing: error)
                }
            }
        }

        public func responses(_ request: ResponsesRequest) -> AsyncThrowingStream<Event, Error> {
            let native = Node.mapResponsesRequest(request)
            return AsyncThrowingStream { continuation in
                do {
                    let bridge = EventStreamBridge(continuation: continuation) { [handle] requestId in
                        try? handle.cancel(requestId: requestId)
                    }
                    let requestId = try handle.responses(request: native, listener: bridge)
                    bridge.activate(requestId: requestId)
                } catch {
                    continuation.finish(throwing: error)
                }
            }
        }

        public func cancel(_ requestId: RequestId) async throws {
            let handle = self.handle
            try await runBlocking {
                try handle.cancel(requestId: requestId.value)
            }
        }
    }

    public final class Models: @unchecked Sendable {
        private let handle: MeshNodeHandle

        fileprivate init(handle: MeshNodeHandle) {
            self.handle = handle
        }

        public func recommended() async throws -> [ModelSummary] {
            let handle = self.handle
            return try await runBlocking { try handle.recommendedModels() }
        }

        public func search(_ query: ModelSearchQuery) async throws -> [ModelSummary] {
            let handle = self.handle
            return try await runBlocking { try handle.searchModels(query: query) }
        }

        public func show(_ modelRef: String) async throws -> ModelDetails {
            let handle = self.handle
            return try await runBlocking { try handle.showModel(modelRef: modelRef) }
        }

        public func installed() async throws -> [InstalledModel] {
            let handle = self.handle
            return try await runBlocking { try handle.installedModels() }
        }

        public func cacheStatus() async throws -> ModelCacheStatus {
            let handle = self.handle
            return try await runBlocking { try handle.modelCacheStatus() }
        }

        public func download(_ modelRef: String) async throws -> DownloadedModel {
            let handle = self.handle
            return try await runBlocking { try handle.downloadModel(modelRef: modelRef) }
        }

        public func delete(_ modelRef: String, options: DeleteModelOptions) async throws -> DeleteModelResult {
            let handle = self.handle
            return try await runBlocking { try handle.deleteModel(modelRef: modelRef, options: options) }
        }

        public func cleanup(_ policy: CleanupPolicy) async throws -> CleanupResult {
            let handle = self.handle
            return try await runBlocking { try handle.cleanupModels(policy: policy) }
        }

        public func pruneDerivedCache(_ policy: PrunePolicy) async throws -> PruneResult {
            let handle = self.handle
            return try await runBlocking { try handle.pruneDerivedCache(policy: policy) }
        }
    }

    public final class Serving: @unchecked Sendable {
        private let handle: MeshNodeHandle

        fileprivate init(handle: MeshNodeHandle) {
            self.handle = handle
        }

        public func status() async throws -> ServingStatus {
            let handle = self.handle
            return try await runBlocking { try handle.servingStatus() }
        }

        public func servedModels() async throws -> [ServedModel] {
            let handle = self.handle
            return try await runBlocking { try handle.servedModels() }
        }

        public func load(_ modelRef: String, options: LoadModelOptions) async throws -> ServedModel {
            let handle = self.handle
            return try await runBlocking { try handle.loadServingModel(modelRef: modelRef, options: options) }
        }

        public func unload(_ target: UnloadTarget, options: UnloadModelOptions) async throws {
            let handle = self.handle
            return try await runBlocking { try handle.unloadServingModel(target: target, options: options) }
        }

        public func unloadModel(_ modelId: String, options: UnloadModelOptions) async throws {
            let handle = self.handle
            return try await runBlocking { try handle.unloadServingModelById(modelId: modelId, options: options) }
        }

        public func unloadInstance(_ instanceId: String, options: UnloadModelOptions) async throws {
            let handle = self.handle
            return try await runBlocking { try handle.unloadServingInstance(instanceId: instanceId, options: options) }
        }

        public func setDevicePolicy(_ policy: DevicePolicy) async throws {
            let handle = self.handle
            return try await runBlocking { try handle.setDevicePolicy(policy: policy) }
        }
    }

    fileprivate static func mapModel(_ native: ModelNative) -> Model {
        Model(id: native.id, name: native.name)
    }

    static func mapEvent(_ native: ClientEvent) -> Event {
        switch native {
        case .connecting:
            return .connecting
        case .joined(let nodeId):
            return .joined(nodeId: nodeId)
        case .modelsUpdated(let models):
            return .modelsUpdated(models: models.map(mapModel))
        case .tokenDelta(let requestId, let delta):
            return .tokenDelta(requestId: requestId, delta: delta)
        case .completed(let requestId):
            return .completed(requestId: requestId)
        case .failed(let requestId, let error):
            return .failed(requestId: requestId, error: error)
        case .disconnected(let reason):
            return .disconnected(reason: reason)
        }
    }

    private static func mapChatRequest(_ request: ChatRequest) -> ChatRequestNative {
        ChatRequestNative(
            model: request.model,
            messages: request.messages.map {
                ChatMessageNative(role: $0.role, content: $0.content)
            }
        )
    }

    private static func mapResponsesRequest(_ request: ResponsesRequest) -> ResponsesRequestNative {
        ResponsesRequestNative(model: request.model, input: request.input)
    }
}
#elseif MESH_SWIFT_STUB
public func generateOwnerKeypairHex() -> String {
    UUID().uuidString.replacingOccurrences(of: "-", with: "").lowercased()
}

public struct PublicMeshQuery: Sendable, Hashable {
    public var model: String?
    public var minVramGb: Double?
    public var region: String?
    public var targetName: String?
    public var relays: [String]

    public init(model: String?, minVramGb: Double?, region: String?, targetName: String?, relays: [String]) {
        self.model = model
        self.minVramGb = minVramGb
        self.region = region
        self.targetName = targetName
        self.relays = relays
    }
}

public struct PublicMesh: Sendable, Hashable {
    public var inviteToken: String
}

public enum CapabilityLevel: Sendable, Hashable {
    case none
    case likely
    case supported
}

public struct ModelCapabilities: Sendable, Hashable {
    public var multimodal: Bool
    public var vision: CapabilityLevel
    public var audio: CapabilityLevel
    public var reasoning: CapabilityLevel
    public var toolUse: CapabilityLevel
    public var moe: Bool
}

public struct ModelSummary: Sendable, Hashable {
    public var id: String
    public var name: String
    public var sizeLabel: String?
    public var description: String?
    public var capabilities: ModelCapabilities
}

public struct ModelSearchQuery: Sendable, Hashable {
    public var query: String
    public var limit: UInt64?

    public init(query: String, limit: UInt64?) {
        self.query = query
        self.limit = limit
    }
}

public enum ModelSource: Sendable, Hashable {
    case catalog
    case huggingFace
    case local
}

public enum ModelKind: Sendable, Hashable {
    case gguf
    case safetensors
    case layerPackage
    case unknown
}

public struct ModelDetails: Sendable, Hashable {
    public var id: String
    public var name: String
    public var source: ModelSource
    public var kind: ModelKind
    public var modelRef: String
    public var downloadRef: String
    public var path: String?
    public var sizeBytes: UInt64?
    public var sizeLabel: String?
    public var description: String?
    public var draft: String?
    public var installed: Bool
    public var capabilities: ModelCapabilities
}

public struct InstalledModel: Sendable, Hashable {
    public var modelRef: String
    public var path: String
    public var sizeBytes: UInt64?
    public var capabilities: ModelCapabilities
}

public struct ModelCacheStatus: Sendable, Hashable {
    public var cacheDir: String?
}

public struct DownloadedModel: Sendable, Hashable {
    public var modelRef: String
    public var paths: [String]
    public var primaryPath: String?
    public var details: ModelDetails?
}

public struct DeleteModelOptions: Sendable, Hashable {
    public var force: Bool

    public init(force: Bool) {
        self.force = force
    }
}

public struct DeleteModelResult: Sendable, Hashable {
    public var deletedPaths: [String]
    public var reclaimedBytes: UInt64
}

public struct CleanupPolicy: Sendable, Hashable {
    public var removeAll: Bool

    public init(removeAll: Bool) {
        self.removeAll = removeAll
    }
}

public struct CleanupResult: Sendable, Hashable {
    public var deletedPaths: [String]
    public var reclaimedBytes: UInt64
    public var skippedPaths: [String]
}

public struct PrunePolicy: Sendable, Hashable {
    public var removeAll: Bool

    public init(removeAll: Bool) {
        self.removeAll = removeAll
    }
}

public struct PruneResult: Sendable, Hashable {
    public var deletedPaths: [String]
    public var reclaimedBytes: UInt64
}

public enum DevicePolicy: Sendable, Hashable {
    case auto
    case cpu
    case gpu(deviceIds: [String])
}

public struct LoadModelOptions: Sendable, Hashable {
    public var devicePolicy: DevicePolicy

    public init(devicePolicy: DevicePolicy) {
        self.devicePolicy = devicePolicy
    }
}

public enum ServingModelState: Sendable, Hashable {
    case loading
    case ready
    case failed
    case unloading
    case stopped
    case unknown(value: String)
}

public struct ServedModel: Sendable, Hashable {
    public var modelRef: String
    public var modelId: String
    public var instanceId: String?
    public var state: ServingModelState
    public var backend: String?
    public var capabilities: ModelCapabilities
    public var contextLength: UInt32?
    public var error: String?
}

public struct ServingStatus: Sendable, Hashable {
    public var enabled: Bool
    public var models: [ServedModel]
}

public enum UnloadTarget: Sendable, Hashable {
    case model(modelId: String)
    case instance(instanceId: String)
}

public struct UnloadModelOptions: Sendable, Hashable {
    public var drainTimeoutMs: UInt64
    public var force: Bool

    public init(drainTimeoutMs: UInt64, force: Bool) {
        self.drainTimeoutMs = drainTimeoutMs
        self.force = force
    }
}

public final class Node: @unchecked Sendable {
    public let inference: Inference
    public let models: Models
    public let serving: Serving
    private var isConnected: Bool = false

    public init(
        inviteToken _: InviteToken,
        ownerKeypairBytesHex _: String,
        cacheDir _: String? = nil,
        runtimeDir _: String? = nil,
        servingEnabled _: Bool = false
    ) throws {
        self.inference = Inference()
        self.models = Models()
        self.serving = Serving()
    }

    public static func discoverPublicMeshes(
        _ query: PublicMeshQuery = PublicMeshQuery(
            model: nil,
            minVramGb: nil,
            region: nil,
            targetName: nil,
            relays: []
        )
    ) async throws -> [PublicMesh] {
        []
    }

    public static func connectPublic(
        ownerKeypairBytesHex _: String,
        query _: PublicMeshQuery = PublicMeshQuery(
            model: nil,
            minVramGb: nil,
            region: nil,
            targetName: nil,
            relays: []
        )
    ) async throws -> Node {
        try Node(inviteToken: InviteToken(""), ownerKeypairBytesHex: "")
    }

    public func start() async throws {
        isConnected = true
    }

    public func stop() async throws {
        isConnected = false
    }

    public func reconnect() async throws {
        isConnected = true
    }

    public func status() async -> Status {
        Status(connected: isConnected, peerCount: 0)
    }

    public final class Inference: @unchecked Sendable {
        public func listModels() async throws -> [Model] {
            []
        }

        public func chat(_ request: ChatRequest) -> AsyncThrowingStream<Event, Error> {
            let requestId = UUID().uuidString
            return AsyncThrowingStream { continuation in
                continuation.yield(.completed(requestId: requestId))
                continuation.finish()
            }
        }

        public func responses(_ request: ResponsesRequest) -> AsyncThrowingStream<Event, Error> {
            let requestId = UUID().uuidString
            return AsyncThrowingStream { continuation in
                continuation.yield(.completed(requestId: requestId))
                continuation.finish()
            }
        }

        public func cancel(_ requestId: RequestId) async throws {}
    }

    public final class Models: @unchecked Sendable {
        public func recommended() async throws -> [ModelSummary] { [] }
        public func search(_ query: ModelSearchQuery) async throws -> [ModelSummary] { [] }
        public func show(_ modelRef: String) async throws -> ModelDetails {
            ModelDetails(
                id: modelRef,
                name: modelRef,
                source: .local,
                kind: .unknown,
                modelRef: modelRef,
                downloadRef: modelRef,
                path: nil,
                sizeBytes: nil,
                sizeLabel: nil,
                description: nil,
                draft: nil,
                installed: false,
                capabilities: Self.emptyCapabilities
            )
        }
        public func installed() async throws -> [InstalledModel] { [] }
        public func cacheStatus() async throws -> ModelCacheStatus { ModelCacheStatus(cacheDir: nil) }
        public func download(_ modelRef: String) async throws -> DownloadedModel {
            DownloadedModel(modelRef: modelRef, paths: [], primaryPath: nil, details: nil)
        }
        public func delete(_ modelRef: String, options: DeleteModelOptions) async throws -> DeleteModelResult {
            DeleteModelResult(deletedPaths: [], reclaimedBytes: 0)
        }
        public func cleanup(_ policy: CleanupPolicy) async throws -> CleanupResult {
            CleanupResult(deletedPaths: [], reclaimedBytes: 0, skippedPaths: [])
        }
        public func pruneDerivedCache(_ policy: PrunePolicy) async throws -> PruneResult {
            PruneResult(deletedPaths: [], reclaimedBytes: 0)
        }

        private static let emptyCapabilities = ModelCapabilities(
            multimodal: false,
            vision: .none,
            audio: .none,
            reasoning: .none,
            toolUse: .none,
            moe: false
        )
    }

    public final class Serving: @unchecked Sendable {
        public func status() async throws -> ServingStatus { ServingStatus(enabled: false, models: []) }
        public func servedModels() async throws -> [ServedModel] { [] }
        public func load(_ modelRef: String, options: LoadModelOptions) async throws -> ServedModel {
            ServedModel(
                modelRef: modelRef,
                modelId: modelRef,
                instanceId: nil,
                state: .stopped,
                backend: nil,
                capabilities: ModelCapabilities(
                    multimodal: false,
                    vision: .none,
                    audio: .none,
                    reasoning: .none,
                    toolUse: .none,
                    moe: false
                ),
                contextLength: nil,
                error: nil
            )
        }
        public func unload(_ target: UnloadTarget, options: UnloadModelOptions) async throws {}
        public func unloadModel(_ modelId: String, options: UnloadModelOptions) async throws {}
        public func unloadInstance(_ instanceId: String, options: UnloadModelOptions) async throws {}
        public func setDevicePolicy(_ policy: DevicePolicy) async throws {}
    }
}
#else
#error("MeshLLM Swift SDK requires MeshLLMFFI.xcframework. Build it with sdk/swift/scripts/build-xcframework.sh or set MESH_SWIFT_FORCE_STUB=1 for unit tests only.")
#endif

private func runBlocking<T>(_ work: @escaping () throws -> T) async throws -> T {
    try await withCheckedThrowingContinuation { continuation in
        DispatchQueue.global(qos: .userInitiated).async {
            do {
                continuation.resume(returning: try work())
            } catch {
                continuation.resume(throwing: error)
            }
        }
    }
}

private func runBlocking<T>(_ work: @escaping () -> T) async -> T {
    await withCheckedContinuation { continuation in
        DispatchQueue.global(qos: .userInitiated).async {
            continuation.resume(returning: work())
        }
    }
}
